WalletWasabi.Tests/UnitTests/Crypto/ZeroKnowledge/LinearRelationTests.cs

```diff
@@ -0,0 +1,109 @@
+using NBitcoin.Secp256k1;
+using System.Linq;
+using WalletWasabi.Crypto;
+using WalletWasabi.Crypto.Groups;
+using WalletWasabi.Crypto.ZeroKnowledge.LinearRelation;
+using Xunit;
+
+namespace WalletWasabi.Tests.UnitTests.Crypto.ZeroKnowledge
+{
+	public class EquationTest
+	{
+		[Theory]
+		[InlineData(0, 0)]
+		[InlineData(0, 1)]
+		[InlineData(1, 0)]
+		[InlineData(1, 1)]
+		[InlineData(1, 2)]
+		[InlineData(3, 5)]
+		[InlineData(5, 7)]
+		[InlineData(7, 11)]
+		[InlineData(short.MaxValue, uint.MaxValue)]
+		[InlineData(int.MaxValue, uint.MaxValue)]
+		[InlineData(uint.MaxValue, uint.MaxValue)]
+		public void VerifyResponsesAndSimulations(uint scalarSeed1, uint scalarSeed2)
+		{
+			var witness = new ScalarVector(new[] { new Scalar(scalarSeed1), new Scalar(scalarSeed2) });
+			var generators = new GroupElementVector(new[] { Generators.G, Generators.Ga });
+			var publicPoint = witness * generators;
+
+			var eqn = new Equation(publicPoint, generators);
+
+			// First, demonstrate proving knowledge with the witness
+			var secretNonces = new ScalarVector(new[] { new Scalar(23), new Scalar(42) });
+			var publicNonce = Enumerable.Zip(secretNonces, generators, (s, g) => s * g).Sum();
+			var challenge = new Scalar(101);
+			var response = eqn.Respond(witness, secretNonces, challenge);
```

Comment:
I would consider using `var`s less in this type of code. Explicit types are easier to reason about and one cannot make a mistake about type.

Reply:
Given this code could evolve soon I would prefer `var` here (specially here in tests) because it makes the maintenance much painless.